spec/_attachments/ic.did

```diff
@@ -2,115 +2,142 @@ type canister_id = principal;
 };
 
 type definite_canister_settings = record {
-  controllers : vec principal;
-  compute_allocation : nat;
-  memory_allocation : nat;
-  freezing_threshold : nat;
-  reserved_cycles_limit : nat;
+    controllers : vec principal;
+    compute_allocation : nat;
+    memory_allocation : nat;
+    freezing_threshold : nat;
+    reserved_cycles_limit : nat;
 };
 
 type change_origin = variant {
-  from_user : record {
-    user_id : principal;
-  };
-  from_canister : record {
-    canister_id : principal;
-    canister_version : opt nat64;
-  };
+    from_user : record {
+        user_id : principal;
+    };
+    from_canister : record {
+        canister_id : principal;
+        canister_version : opt nat64;
+    };
 };
 
 type change_details = variant {
-  creation : record {
-    controllers : vec principal;
-  };
-  code_uninstall;
-  code_deployment : record {
-    mode : variant {install; reinstall; upgrade};
-    module_hash : blob;
-  };
-  controllers_change : record {
-    controllers : vec principal;
-  };
+    creation : record {
+        controllers : vec principal;
+    };
+    code_uninstall;
+    code_deployment : record {
+        mode : variant { install; reinstall; upgrade };
+        module_hash : blob;
+    };
+    controllers_change : record {
+        controllers : vec principal;
+    };
 };
 
 type change = record {
-  timestamp_nanos : nat64;
-  canister_version : nat64;
-  origin : change_origin;
-  details : change_details;
+    timestamp_nanos : nat64;
+    canister_version : nat64;
+    origin : change_origin;
+    details : change_details;
 };
 
 type chunk_hash = blob;
 
-type http_header = record { name: text; value: text };
+type http_header = record {
+    name : text;
+    value : text;
+};
 
-type http_response = record {
-  status: nat;
-  headers: vec http_header;
-  body: blob;
+type http_request_result = record {
+    status : nat;
+    headers : vec http_header;
+    body : blob;
 };
 
-type ecdsa_curve = variant { secp256k1; };
+type ecdsa_curve = variant {
+    secp256k1;
+};
 
 type satoshi = nat64;
 
 type bitcoin_network = variant {
-  mainnet;
-  testnet;
+    mainnet;
+    testnet;
 };
 
 type bitcoin_address = text;
 
 type block_hash = blob;
 
 type outpoint = record {
-  txid : blob;
-  vout : nat32
+    txid : blob;
+    vout : nat32;
 };
 
 type utxo = record {
-  outpoint: outpoint;
-  value: satoshi;
-  height: nat32;
+    outpoint : outpoint;
+    value : satoshi;
+    height : nat32;
 };
 
-type get_utxos_request = record {
-  address : bitcoin_address;
-  network: bitcoin_network;
-  filter: opt variant {
-    min_confirmations: nat32;
-    page: blob;
-  };
+type bitcoin_get_utxos_args = record {
+    address : bitcoin_address;
+    network : bitcoin_network;
+    filter : opt variant {
+        min_confirmations : nat32;
+        page : blob;
+    };
 };
 
-type get_current_fee_percentiles_request = record {
-  network: bitcoin_network;
+type bitcoin_get_utxos_query_args = record {
```

Comment:
Do we need to duplicate the type for query methods since they're exactly the same as for update methods?

Reply:
i think  you're right and the chance that those two types diverge is probably 0, but for the sake of consistency i'd just keep this duplicate

Reply:
What's the consistency reference? Is the source code of those really duplicated?

Reply:
> What's the consistency reference? Is the source code of those really duplicated?

that the parameters of each method have their own type definition named `<method_name>_args`.

Reply:
Yeah ok but, is it so in the source code?

Reply:
I don't know, why do you think that matters?

Reply:
I case you haven't seen it, [this](https://github.com/dfinity/interface-spec/issues/269) was the motivation.

Reply:
Because the did files should reflect the implementation not a pattern. So if in the source code both parameters are actually inheriting the same type, those should not be duplicated in the did file regarless of the "consistency" pattern. Therefore the source code matters.

Ultimately I'm also curious about this because it implies that I should duplicate code in ic-js.

Reply:
It's the same type in the source code. See [implementation](https://github.com/dfinity/ic/blob/5fe907da2193c6051634b29133280f53f2490d52/rs/tests/src/execution/queries.rs#L136).

Therefore I think it's incorrect to duplicate the types in the specification. Agree?

Reply:
> Because the did files should reflect the implementation not a pattern.

i'd argue that the interface and the implementation can be viewed separately. e.g. i could have a public method that i don't want to add to my `did` file. but yes, in any case they should be in sync but i don't think that means they have to mirror each other regarding their structure. (this is a lightly held opinion :D)

Reply:
I've implemented those two in this ic-js PR now awyway. https://github.com/dfinity/ic-js/pull/559
Thanks for the feedback.